---
title: "Objects"
output: html_document
date: "2023-04-06"
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
getwd()
```

```{r}
library(rtracklayer)
library(magrittr)
library(dplyr)
library(magrittr)
library(data.table)
```

```{r}
gtf0 <- import("GCF_027941655.1_ASM2794165v1_genomic.gtf.gz")
gtf0
```

```{r}
names(mcols(gtf0)) 
```

```{r}
str(gtf0)
```

```{r}
methods(class = GRanges)
as.data.frame(gtf0) %>% head()
```

```{r}
temp <- mcols(gtf0)
temp <- as.data.frame(temp)
temp
```

```{r}
head(gtf0$type)   
levels(gtf0$type) 
table(gtf0$type)
table(gtf0$gene_biotype)
```

```{r}
gtf1 <- gtf0[gtf0$type == "exon"|gtf0$type == "CDS"]
gtf1
```

```{r}
mcols(gtf1)
names(mcols(gtf1))
```

```{r}
gtf1
table(as.data.frame(gtf1$gbkey))
```

```{r}
is.na(gtf1$transcript_id) %>% head()
is.na(gtf1$transcript_id) %>% tail()
```

```{r}
is.na(gtf1$transcript_id) %>% sum()
```

```{r}
gtf2 <- gtf1[!is.na(gtf1$transcript_id)  ]
gtf2
```

```{r}
gtf2$transcript_id %>% head()
duplicated(gtf2$transcript_id) %>% head()
```

```{r}
as.data.frame(gtf2) %>% head()
names(mcols(gtf2)) 
```

```{r}
gtf3  <- as.data.frame(gtf2)[,c("transcript_id","gene_id","gene","product","protein_id")]
head(gtf3)
```

```{r}
gtf3$protein_id <- ifelse(is.na(gtf3$protein_id), gtf3$gene_id, gtf3$protein_id)
head(gtf3)
```

```{r}
write.csv(gtf3, 
          file = "GCF_027941655.1_ASM2794165v1.csv",
          row.names = FALSE)
```

```{r}
sessionInfo()
```


